app/channel_test.go

```diff
@@ -1041,3 +1041,78 @@ func TestSearchChannelsForUser(t *testing.T) {
 		searchAndCheck(t, "dev", []string{"test-dev-1", "test-dev-2", "dev-3"})
 	})
 }
+
+func TestMarkChannelAsUnreadFromPost(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	u1 := th.BasicUser
+	u2 := th.BasicUser2
+	c1 := th.BasicChannel
+	pc1 := th.CreatePrivateChannel(th.BasicTeam)
+	th.AddUserToChannel(u2, c1)
+	th.AddUserToChannel(u1, pc1)
+	th.AddUserToChannel(u2, pc1)
+
+	_ = th.CreatePost(c1)
+	p2 := th.CreatePost(c1)
+	_ = th.CreatePost(c1)
+
+	pp1 := th.CreatePost(pc1)
+	require.NotNil(t, pp1)
+	pp2 := th.CreatePost(pc1)
+
+	t.Run("Ensure that post have been read", func(t *testing.T) {
+		unread, err := th.App.GetChannelUnread(c1.Id, u1.Id)
+		require.Nil(t, err)
+		require.Equal(t, int64(4), unread.MsgCount)
+		unread, err = th.App.GetChannelUnread(c1.Id, u2.Id)
+		require.Nil(t, err)
+		require.Equal(t, int64(4), unread.MsgCount)
+		th.App.UpdateChannelLastViewedAt([]string{c1.Id, pc1.Id}, u1.Id)
+		th.App.UpdateChannelLastViewedAt([]string{c1.Id, pc1.Id}, u2.Id)
+		require.Nil(t, err)
+		unread, err = th.App.GetChannelUnread(c1.Id, u2.Id)
+		require.Nil(t, err)
+		require.Equal(t, int64(0), unread.MsgCount)
+	})
+
+	t.Run("Unread last one", func(t *testing.T) {
+		response, err := th.App.MarkChannelAsUnreadFromPost(p2.Id, u1.Id)
+		require.Nil(t, err)
+		require.NotNil(t, response)
+		assert.Equal(t, int64(1), response.MsgCount)
+		unread, err := th.App.GetChannelUnread(c1.Id, u1.Id)
+		require.Nil(t, err)
+		assert.Equal(t, int64(1), unread.MsgCount)
+	})
+
+	t.Run("Other users are unaffected", func(t *testing.T) {
+		unread, err := th.App.GetChannelUnread(c1.Id, u2.Id)
+		require.Nil(t, err)
+		assert.Equal(t, int64(0), unread.MsgCount)
+
+	})
+
+	t.Run("Unread on a private channel", func(t *testing.T) {
```

Comment:
Do we need a separate test for private channels? It doesn't seem like this behaves any differently between public and private channels

Reply:
I'm usually favouring this kind of test for future changes between private and public, to ensure some future change doesn't affect functionality.